azurerm_cognitive_account supports identity storage disable_local_auth fqdns public_network_access_enabled and restrict_outbound_network_access

ms-henglu opened the pull request:
`azurerm_cognitive_account` supports `identity` `storage` `disable_local_auth` `fqdns` `public_network_access_enabled` and `restrict_outbound_network_access`

The tests are listed as the followings.

```

=== RUN   TestAccCognitiveAccount_basic
=== PAUSE TestAccCognitiveAccount_basic
=== CONT  TestAccCognitiveAccount_basic
--- PASS: TestAccCognitiveAccount_basic (178.22s)
=== RUN   TestAccCognitiveAccount_speechServices
=== PAUSE TestAccCognitiveAccount_speechServices
=== CONT  TestAccCognitiveAccount_speechServices
--- PASS: TestAccCognitiveAccount_speechServices (164.60s)
=== RUN   TestAccCognitiveAccount_speechServicesWithStorage
=== PAUSE TestAccCognitiveAccount_speechServicesWithStorage
=== CONT  TestAccCognitiveAccount_speechServicesWithStorage
--- PASS: TestAccCognitiveAccount_speechServicesWithStorage (207.76s)
=== RUN   TestAccCognitiveAccount_requiresImport
=== PAUSE TestAccCognitiveAccount_requiresImport
=== CONT  TestAccCognitiveAccount_requiresImport
--- PASS: TestAccCognitiveAccount_requiresImport (184.34s)
=== RUN   TestAccCognitiveAccount_complete
=== PAUSE TestAccCognitiveAccount_complete
=== CONT  TestAccCognitiveAccount_complete
--- PASS: TestAccCognitiveAccount_complete (181.70s)
=== RUN   TestAccCognitiveAccount_update
=== PAUSE TestAccCognitiveAccount_update
=== CONT  TestAccCognitiveAccount_update
--- PASS: TestAccCognitiveAccount_update (257.20s)
=== RUN   TestAccCognitiveAccount_qnaRuntimeEndpoint
=== PAUSE TestAccCognitiveAccount_qnaRuntimeEndpoint
=== CONT  TestAccCognitiveAccount_qnaRuntimeEndpoint
--- PASS: TestAccCognitiveAccount_qnaRuntimeEndpoint (244.40s)
=== RUN   TestAccCognitiveAccount_qnaRuntimeEndpointUnspecified
=== PAUSE TestAccCognitiveAccount_qnaRuntimeEndpointUnspecified
=== CONT  TestAccCognitiveAccount_qnaRuntimeEndpointUnspecified
--- PASS: TestAccCognitiveAccount_qnaRuntimeEndpointUnspecified (91.48s)
=== RUN   TestAccCognitiveAccount_cognitiveServices
=== PAUSE TestAccCognitiveAccount_cognitiveServices
=== CONT  TestAccCognitiveAccount_cognitiveServices
--- PASS: TestAccCognitiveAccount_cognitiveServices (154.87s)
=== RUN   TestAccCognitiveAccount_withMultipleCognitiveAccounts
=== PAUSE TestAccCognitiveAccount_withMultipleCognitiveAccounts
=== CONT  TestAccCognitiveAccount_withMultipleCognitiveAccounts
=== CONT  TestAccCognitiveAccount_withMultipleCognitiveAccounts
--- PASS: TestAccCognitiveAccount_withMultipleCognitiveAccounts (109.40s)
=== RUN   TestAccCognitiveAccount_networkAcls
=== PAUSE TestAccCognitiveAccount_networkAcls
=== CONT  TestAccCognitiveAccount_networkAcls
--- PASS: TestAccCognitiveAccount_networkAcls (345.53s)
=== RUN   TestAccCognitiveAccount_identity
=== PAUSE TestAccCognitiveAccount_identity
=== CONT  TestAccCognitiveAccount_identity
--- PASS: TestAccCognitiveAccount_identity (472.47s)

=== RUN   TestAccCognitiveAccount_metricsAdvisor
=== PAUSE TestAccCognitiveAccount_metricsAdvisor
=== CONT  TestAccCognitiveAccount_metricsAdvisor
--- PASS: TestAccCognitiveAccount_metricsAdvisor (502.44s)

```

katbyte commented on azurerm/internal/services/cognitive/cognitive_account_resource.go:
```diff
@@ -104,6 +108,83 @@ func resourceCognitiveAccount() *pluginsdk.Resource {
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+
+			"disable_local_auth": {
```
could we change this to be more consistant with the provider
```suggestion
			"local_auth_enabled": {
```

katbyte commented on azurerm/internal/services/cognitive/cognitive_account_resource.go:
```diff
@@ -104,6 +108,83 @@ func resourceCognitiveAccount() *pluginsdk.Resource {
+				Type:         pluginsdk.TypeString,
+				Optional:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
```
could we properly validate this?

katbyte commented on azurerm/internal/services/cognitive/cognitive_account_resource.go:
```diff
@@ -104,6 +108,83 @@ func resourceCognitiveAccount() *pluginsdk.Resource {
+				Type:         pluginsdk.TypeString,
+				Optional:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
```
could we properly validate this?

katbyte commented on azurerm/internal/services/cognitive/cognitive_account_resource.go:
```diff
@@ -104,6 +108,83 @@ func resourceCognitiveAccount() *pluginsdk.Resource {
+				Optional: true,
+				Elem: &pluginsdk.Schema{
+					Type:         pluginsdk.TypeString,
+					ValidateFunc: validation.StringIsNotEmpty,
```
we cann properly check this is a FQDN?

ms-henglu replied:
fqdn is quite complicated to validate, maybe we can skip it?

katbyte replied:
sure

katbyte commented on azurerm/internal/services/cognitive/cognitive_account_resource.go:
```diff
@@ -146,7 +227,46 @@ func resourceCognitiveAccount() *pluginsdk.Resource {
+				Default:  true,
+			},
+
+			"restrict_outbound_network_access": {
```
could we put restrict on the end to be consistant with other bools?
```suggestion
			"outbound_network_access_restrited": {
```

katbyte commented on azurerm/internal/services/cognitive/cognitive_account_resource.go:
```diff
@@ -146,7 +227,46 @@ func resourceCognitiveAccount() *pluginsdk.Resource {
+				Optional: true,
+				Elem: &pluginsdk.Resource{
+					Schema: map[string]*pluginsdk.Schema{
+						"resource_id": {
```
what type of resource is this?

ms-henglu replied:
storage account id, I think it's better to rename it to `storage_account_id`. I'll change that.

katbyte commented on azurerm/internal/services/cognitive/cognitive_account_resource.go:
```diff
@@ -146,7 +227,46 @@ func resourceCognitiveAccount() *pluginsdk.Resource {
+				},
+			},
+
+			"super_user": {
```
is this user name?
```suggestion
			"super_user_name": {
```

katbyte commented on website/docs/r/cognitive_account.html.markdown:
```diff
@@ -56,6 +60,22 @@ The following arguments are supported:
+
+* `storage` - (Optional) An `identity` block is documented below.
+
+* `super_user` - (Optional) The super user of Metrics Advisor. This attribute is only set when kind is `MetricsAdvisor`. Changing this forces a new resource to be created.
```
should this be
```suggestion
* `metrics_advisor_super_user_name` - (Optional) The super user of Metrics Advisor. This attribute is only set when kind is `MetricsAdvisor`. Changing this forces a new resource to be created.
```

katbyte commented on website/docs/r/cognitive_account.html.markdown:
```diff
@@ -44,10 +44,14 @@ The following arguments are supported:
 
 * `sku_name` - (Required) Specifies the SKU Name for this Cognitive Service Account. Possible values are `F0`, `F1`, `S`, `S0`, `S1`, `S2`, `S3`, `S4`, `S5`, `S6`, `P0`, `P1`, and `P2`.
 
+* `aad_client_id` - (Optional) The Azure AD Client ID (Application ID). This attribute is only set when kind is `MetricsAdvisor`. Changing this forces a new resource to be created.
```
could we validate that when kind is MetricsAdvisor this is set? also given this is just for mstrics advisor:
```suggestion
* `metrics_advisor_aad_client_id` - (Optional) The Azure AD Client ID (Application ID). This attribute is only set when kind is `MetricsAdvisor`. Changing this forces a new resource to be created.
```

ms-henglu replied:
It's not required for metrics advisor, so I think it's not necessary to validate that when kind is MetricsAdvisor this is set.

katbyte reviewed (changes requested):
Thanks @ms-henglu - i've left some comments inline. we also have some failing tests:
![image](https://user-images.githubusercontent.com/1638467/124647733-35b68680-de4b-11eb-84fa-af393afda210.png)

ms-henglu commented:
@katbyte , Hi , I've added a commit to rename these fields. And about the failed test, the cause is that QnAMaker and SpeechServices are not available in `West Europe`, so I run the tests again in `West US`.
![image](https://user-images.githubusercontent.com/79895375/124699383-48998b80-df1d-11eb-9317-2df0af4684fb.png)

katbyte commented:
@ms-henglu - that is not going to fix it for the nightly runs. is it availible in any of these regions?
![image](https://user-images.githubusercontent.com/1638467/124943170-20fbff00-dfc1-11eb-887c-458f09040f8a.png)

ms-henglu commented:
@katbyte , for `SpeechServices`, it's available in `eastus2` and `francecentral`, so I fixed it. But `QnAMaker` is only available in `west us`, so I hardcoded it in tests and added comments for it.

Only one test failed in tc, because it needs to register a subscription through[ this form](https://forms.office.com/Pages/ResponsePage.aspx?id=v4j5cvGGr0GRqy180BHbR3k4-f_9d9xPhWkgOUub9YhUNDZJUERRVVVDME4xNDVNRjEwMTNZV1dHSS4u). Could you please help me with it? Because I can't see the subscriptionId.

tc tests.
![image](https://user-images.githubusercontent.com/79895375/125011778-a9e76900-e09b-11eb-90d3-71d264d26a1e.png)

local tests with same paramters.
![image](https://user-images.githubusercontent.com/79895375/125011990-fe8ae400-e09b-11eb-9f64-2b1e8925e99e.png)

katbyte reviewed (approved):
Thanks @ms-henglu - all tests pass and this LGTM 🙌
